### ISSUE TITLE
Forgotten completion call

### ISSUE THREAD

**Issue by ekazaev** [GetStream/stream-chat-swift#2545]
## What did you do?
`loadNextMessages` missing a completion call. In case it has all loaded messages comletion will never be caller. Incorrect situation. Completion should be called always with an error or a result.

```swift
   public func loadNextMessages(
        after messageId: MessageId? = nil,
        limit: Int? = nil,
        completion: ((Error?) -> Void)? = nil
    )

        guard !hasLoadedAllNextMessages && !isLoadingNextMessages else {
            log.debug("""
                Skipped loading more next messages.
                    - hasLoadedAllNextMessages:\(hasLoadedAllNextMessages)
                    - isLoadingNextMessages: \(isLoadingNextMessages)
            """)
            return
        }
```
## GetStream Environment
**GetStream Chat version:** 4.29.0 
**GetStream Chat frameworks:** StreamChat

**Comment by nuno-vieira** [comment]
Hi @ekazaev!

Thank you for the report. This was actually done on purpose since if the messages are loaded, we don't call anything, so nothing is completed in theory. What issue is this causing you? We can change this, of course, but we want to understand what is the use case so that we don't miss anything. 👍 

Best,
Nuno

**Comment by ekazaev** [comment]
@nuno-vieira hi 

First of all it is a break of contract. Imagine you are calling some function with a completion and the completion is never being called back. May I am as the library user want to put some flag back to true or false and so on. To keep the contract you should either call completion all the time or assert it if according to your views it is an incorrect state of the controller that the function was called from. As for me as a user it isn't obvious that i cant call load next/prev methods at some states and i had to go to debug and just copy and paste your code.

**Comment by nuno-vieira** [comment]
Hi @ekazaev,

The idea behind this was to remove as much as possible of the boilerplate code required to load prev/next pages. This is especially useful if a customer is using the UI SDK. But if you are not using it, you are right that this could be unexpected. 

If you have already loaded all messages, you should check the flag `hasLoadedAllPreviousMessages` and avoid calling the function again. 

> Imagine you are calling some function with completion, and the completion is never being called back.

Although I understand what you mean, it could also be unexpected to call the completion if the controller is already loading some data. Right now, we don't call the completion block in 2 scenarios, when the controller has loaded all messages and when the controller is currently loading messages. Imaging the following example:

Example:
```swift
loadingSpinner.isHidden = false
controller.loadNextMessages {
    loadingSpinner.isHidden = true
}
```
In this scenario, if we call the completion when we are in the `isLoadingNextMessages` state, the spinner in the above snippet will be hidden when it shouldn't, which could be unexpected. 

So something like this would be required:
```swift
if controller.hasLoadedAllNextMessages {
    return
}

loadingSpinner.isHidden = false
if controller.isLoadingNextMessages {
     return
}

controller.loadNextMessages {
    loadingSpinner.isHidden = true
}
```

Either way, it does make sense to still call the completion blocks, since you would need to handle this logic either way. Although we provide the state out of the box to the customer, it makes sense for the customer to customise it, or change it at will. So I guess we will change it and call the completion block in both scenarios. 👍

**Comment by ekazaev** [comment]
Hi @nuno-vieira 

boilerplate is boilerplate, breaking the contract is another thing. Then add an assertion that the users of your library should not call this method in that state at all so they are aware about it.

To add to this is another thing.  Lets say I got a pushnotification for some message id. Lets say the one before the last one. I tap on it and start to load messages around it. `hasLoadedAllNextMessages` is incorrectly set to false if I am loading messages around last message id. And there is no way for me to check if I actually have more messages or not.  So from your example - `loadingSpinner` never disappears.

**Comment by alteral** [comment]
Hi @ekazaev,

This issue is now fixed in our latest release, `4.30.0`.
I'm closing this ticket for now, but feel free to reopen it if the provided fix does not solve the issue for you.

Best,
Alexey
